src/lib/transition_frontier_persistence/transition_frontier_persistence.ml

```diff
@@ -106,19 +53,180 @@ end = struct
     Or_error.return x
 
+  module Batch = struct
+    let set_transition batch
+        {With_hash.hash= state_hash; data= external_transition} =
+      Transition_storage.Batch.set batch ~key:state_hash
+        ~data:(External_transition.of_verified external_transition)
+
+    let remove_transition batch state_hash =
+      Transition_storage.Batch.remove batch ~key:state_hash
+  end
+
   let parent_hash transition =
     let open External_transition.Verified in
     let protocol_state = protocol_state transition in
     External_transition.Protocol_state.(previous_state_hash protocol_state)
 
+  module Frontier_diff = struct
+    type add_transition =
+      (External_transition.Verified.t, State_hash.t) With_hash.t
+
+    type move_root =
+      { best_tip: (External_transition.Verified.t, State_hash.t) With_hash.t
+      ; removed_transitions: State_hash.Stable.Latest.t list
+      ; new_root: State_hash.Stable.Latest.t
+      ; new_scan_state: Staged_ledger.Scan_state.Stable.Latest.t }
+
+    type t = Add_transition of add_transition | Move_root of move_root
+  end
+
+  module Diff_mutant = struct
+    type serialized_consensus_state = string
+
+    type move_root =
+      { parent: serialized_consensus_state
+      ; removed_transitions: serialized_consensus_state list
+      ; old_root: State_hash.Stable.Latest.t
+      ; old_scan_state: Staged_ledger.Scan_state.Stable.Latest.t }
+
+    (** Diff_mutant is a GADT that represents operations that affect the
+        changes on the transition_frontier. Only two types of changes can occur
+        when updating the transition_frontier: Add_transition and Move_root.
+        Add_transition would simply add a transition to the frontier with no
+        other side effects. So, the input of Add_transition GADT is an
+        external_transition. To add the transition, we need to know its parent.
+        To certify that we added it to the right parent, we need some
+        representation of the parent. A serialized form of the consensus_state
+        can accomplish this. Therefore, the type of the GADT will be
+        parameterized by a serialized type of the consensus state. The
+        Move_root data type is an operation where we have a new best tip
+        external_transition, remove external_transitions based on their
+        state_hash and update some root data with new root data. Like
+        Add_transition, we can certify that we added the transition into the
+        right parent by showing the serialized_consensus_state of the parent.
+        We can indicate that we removed the external transitions with a certain
+        state_hash by indicating the serialized consensus state of the
+        transition. We can also note which root we are going to replace by
+        indicating the old root *)
+    type _ t =
+      | Add_transition :
+          Frontier_diff.add_transition
+          -> serialized_consensus_state t
+      | Move_root : Frontier_diff.move_root -> move_root t
```

Comment:
Why redefine this twice? Just make the other diff a GADT. If you ever need to talk about it without the type parameter, just wrap it in an existential. This creates an unnecessary disconnection.